### PR TITLE
RHICOMPL-2862: use the ubi8-minimal image as the base

### PR DESCRIPTION
**Pull request by skateman**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,25 +1,50 @@
...
+RUN chown 1001:0 /opt/app-root
 
-RUN /usr/bin/fix-permissions /tmp/src
 USER 1001
 
-# Install the dependencies
-RUN /usr/libexec/s2i/assemble
+ENV PYTHONPATH=/opt/app-root/build/lib PATH="/opt/app-root/bin:$PATH"
+
+RUN python setup.py build
```
Comment (skateman) [inline Dockerfile:41]: This might not be needed, need to check.

File: Dockerfile [hunk 2]
```diff
@@ -1,25 +1,50 @@
...
+
+RUN microdnf install -y python39 libpq
+
+WORKDIR /opt/app-root
+
+COPY --chown=1001:0 --from=build /opt/app-root /opt/app-root
+COPY --chown=1001:0 app.py setup.py /opt/app-root/
+COPY --chown=1001:0 src /opt/app-root/src
+
+RUN chown 1001:0 /opt/app-root
```
[the remark this answers is not in the document]
Comment (skateman) [reply Dockerfile]: It is, in order to the installation + testing pass.

**Comment by skateman** [comment]
@Victoremepunto ready for another round

File: Dockerfile [hunk 3]
```diff
@@ -1,25 +1,47 @@
...
 
 # Add application sources to a directory that the assemble script expects them
 # and set permissions so that the container runs without root access
 USER 0
 
-RUN dnf -y remove httpd-tools httpd-filesystem
+RUN microdnf install -y python39-devel postgresql-devel gcc && \
+    pip3 install virtualenv                                 && \
+    mkdir -p /opt/app-root                                  && \
+    chown 1001:0 /opt/app-root
```
Comment (Victoremepunto) [inline Dockerfile:10]: I think we could keep `python-39` as the *build*  image, it is my understanding there's no usage in prod for it as we use this multi-stage build as we end up using the minimal. 

That way we'd just have to install `postgresql-devel` and `gcc`.

File: Dockerfile [hunk 4]
```diff
@@ -1,25 +1,47 @@
...
-ADD app.py setup.py requirements.txt /tmp/src/
-ADD src /tmp/src/src
+USER 1001
+
+COPY --chown=1001:0 app.py setup.py requirements.txt /opt/app-root/
+COPY --chown=1001:0 src /opt/app-root/src
+
+WORKDIR /opt/app-root
+
+RUN virtualenv .                        && \
```
Comment (Victoremepunto) [inline Dockerfile:19]: We can do this as user 0 or as 1001, but as we're going to copy the files afterwards regarless, I think it's not needed to use the `chown` part of COPY. 

Also, we can first call `virtualenv` to create our directory, then use `WORKDIR` directive to set that virtual env dir. Then, copy files using relative routes, something like 

```
WORKDIR /opt/app-root/src/venv

COPY app.py setup.py requirements.txt .
COPY src ./src
```

File: Dockerfile [hunk 5]
```diff
@@ -1,25 +1,47 @@
...
+
+RUN microdnf install -y python39 libpq && \
+    chown 1001:0 /opt/app-root
 
-RUN /usr/bin/fix-permissions /tmp/src
 USER 1001
 
-# Install the dependencies
-RUN /usr/libexec/s2i/assemble
+ENV PYTHONPATH=/opt/app-root/build/lib PATH="/opt/app-root/bin:$PATH"
```
Comment (Victoremepunto) [inline Dockerfile:38]: I don't understand why we need to specify `PYTHONPATH` , could you please try and remove it ? I believe it should work without it .

File: run-tests.sh [hunk 6]
```diff
@@ -158,18 +158,18 @@ ARTIFACTS_DIR="$WORKSPACE/artifacts"
 mkdir -p "$ARTIFACTS_DIR"
 
 create_env_file || exit 1
-docker cp "$TESTS_ENV_FILE" "$TEST_CONTAINER_ID:/opt/app-root/src/tests/env.yaml"
+docker cp "$TESTS_ENV_FILE" "$TEST_CONTAINER_ID:/opt/app-root/tests/env.yaml"
```
Comment (Victoremepunto) [inline run-tests.sh:161]: If we configure the WORKDIR in `base` I think we can fix this and future problems using relative routes , something like 

```
TEST_CONTAINER_ID:tests/tests/env.yaml
```

File: run-tests.sh [hunk 7]
```diff
@@ -158,18 +158,18 @@ ARTIFACTS_DIR="$WORKSPACE/artifacts"
...
 echo '===================================='
 echo '===     Running Tests           ===='
 echo '===================================='
 set +e
-docker exec "$TEST_CONTAINER_ID" /bin/bash -c "pytest --junitxml=test-report.xml"
+docker exec "$TEST_CONTAINER_ID" /bin/bash -c "pytest --junitxml=test-report.xml tests"
 TEST_RESULT=$?
 set -e
 # Copy test reports
-docker cp "$TEST_CONTAINER_ID:/opt/app-root/src/test-report.xml" "$WORKSPACE/artifacts/junit-test-report.xml"
```
Comment (Victoremepunto) [inline run-tests.sh:172]: same as before, we could use relative routes to avoid future path changes.

**Review by Victoremepunto — COMMENTED** [review]
It looks great, thank you ! let's discuss over some points and then we can merge this.

File: Dockerfile (hunk 3 above)
Comment (skateman) [reply Dockerfile:10]: I am building constantly this on my computer, I don't like the idea of having one extra base image. Also I don't know how the safety scan works with prodsec so I'd like to play it safe.

File: Dockerfile (hunk 4 above)
Comment (skateman) [reply Dockerfile:19]: I don't want to tamper with the WORKDIR if possible, would be better to keep it static. But I'll adjust the paths relatively.

File: run-tests.sh (hunk 7 above)
Comment (Victoremepunto) [reply run-tests.sh:172]: it should definitely work: 

An excerpt of [Docker cd documentation](https://docs.docker.com/engine/reference/commandline/cp/#extended-description):
```
The docker cp command assumes container paths are relative to the container’s / (root) directory. This means supplying the initial forward slash is optional; The command sees compassionate_darwin:/tmp/foo/myfile.txt and compassionate_darwin:tmp/foo/myfile.txt as identical. Local machine paths can be an absolute or relative value. The command interprets a local machine’s relative paths as relative to the current working directory where docker cp is run.
```

What version of Docker cli are you using ? that may be the issue there , but not sure.

Comment (skateman) [reply run-tests.sh:172]: Well, it works for me locally :see_no_evil: show the documentation to Jenkins :laughing:

**Review by Victoremepunto — APPROVED** [review]
Looks good, thanks !!

**Pull request merged**